src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+    // a consistency check. Just whether there is supposed to be an
+    // answer or not.
+    let mut msg = msg.answer()?;
+    for rr in &mut msg {
+        let rr = rr?;
+        if rr.rtype() == qtype && rr.class() == qclass {
+            // We found an answer.
+            return Ok(NoErrorType::Answer);
+        }
+    }
+
+    // No answer. Check the authority section for SOA and NS records.
+    // If the SOA is present then the response is a NODATA response.
+    // If SOA records are absent but NS records are present then the
+    // response is a delegation.
+    let mut found_ns = false;
+    let mut msg = msg.next_section()?.expect("section should be present");
+    for rr in &mut msg {
+        let rr = rr?;
+        if rr.class() == qclass && rr.rtype() == Rtype::Soa {
+            return Ok(NoErrorType::NoData);
+        }
+        if rr.class() == qclass && rr.rtype() == Rtype::Ns {
+            found_ns = true;
+        }
+    }
+
+    if found_ns {
+        return Ok(NoErrorType::Delegation);
+    }
+
+    // Neither SOA nor NS were found. This is a broken response.
+    Ok(NoErrorType::NoErrorWeird)
+}
+
+/// Get the extended rcode of a message.
+fn get_opt_rcode<Octs: Octets>(msg: &Message<Octs>) -> OptRcode {
+    let opt = msg.opt();
+    match opt {
+        Some(opt) => opt.rcode(msg.header()),
+        None => {
+            // Convert Rcode to OptRcode, this should be part of
+            // OptRcode
+            OptRcode::from_int(msg.header().rcode().to_int() as u16)
+        }
+    }
+}
+
+/// Return a message with the AA flag clear.
+fn clear_aa(msg: &Message<Bytes>) -> Result<Message<Bytes>, Error> {
+    // Assume the clear_aa will be called only if AA is set. So no need to
+    // optimize for the case where AA is clear.
+    let mut msg = Message::<Vec<u8>>::from_octets(msg.as_slice().into())?;
+    let hdr = msg.header_mut();
+    hdr.set_aa(false);
+    Ok(Message::<Bytes>::from_octets(msg.into_octets().into())?)
+}
+
+/// Return a message with the AD flag clear.
+fn clear_ad(msg: &Message<Bytes>) -> Result<Message<Bytes>, Error> {
+    // Assume the clear_ad will be called only if AD is set. So no need to
+    // optimize for the case where AD is clear.
+    let mut msg = Message::<Vec<u8>>::from_octets(msg.as_slice().into())?;
+    let hdr = msg.header_mut();
+    hdr.set_ad(false);
+    Ok(Message::<Bytes>::from_octets(msg.into_octets().into())?)
+}
+
+/// Return a message with the RD flag clear.
+fn clear_rd(msg: &Message<Bytes>) -> Result<Message<Bytes>, Error> {
+    // Assume the clear_rd will be called only if RD is set. So no need to
+    // optimize for the case where RD is clear.
+    let mut msg = Message::<Vec<u8>>::from_octets(msg.as_slice().into())?;
+    let hdr = msg.header_mut();
+    hdr.set_rd(false);
+    Ok(Message::<Bytes>::from_octets(msg.into_octets().into())?)
+}
+
+/// Prepare a value for inserting in the cache by clearing the AA flag if
+/// set.
+fn prepare_for_insert<C>(
```

Comment:
This is very similar to repeated code blocks above, only this does `clear_aa()` instead of `clear_rd()` etc. I suggest to refactor these almost common code blocks out (and as said above to use `set_aa(false)` rather than `clear_aa()`).